### PR TITLE
More useful test error messages on should_panic(expected=...) mismatch

### LINKED ISSUE #66304: Better error messages for #[should_panic] tests with expected strings
When running this test ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e6077d5cc62bfc0602db8253f0effbec)):
```rust
#[test]
#[should_panic(expected = "foo\r\n")]
fn panic_test() {
    panic!("foo\n");
}
```

the output is
```
running 1 test
test panic_test ... FAILED

failures:

---- panic_test stdout ----
thread 'panic_test' panicked at 'foo
', src/lib.rs:4:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
note: panic did not include expected string 'foo

'

failures:
    panic_test

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
```

which makes it hard to spot the difference in whitespace characters.

Would it be possible to change this to e.g.:
```
running 1 test
test panic_test ... FAILED

failures:

---- panic_test stdout ----
thread 'panic_test' panicked at 'foo
', src/lib.rs:4:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
note: panic did not include expected string 'foo

'
   Panic message: "foo\n"
 Expected string: "foo\r\n"

failures:
    panic_test

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
```

This would improve the usability when running into these kinds of test failures and also make it more consistent with `assert_eq` which outputs both the left and the right value too.

### PR DESCRIPTION
**Pull request by thomasetter**
Fixes  #66304
r? @gilescope

Shows both the actual as well as the expected panic value when a test with `should_panic(expected=...)` fails.
This makes `should_panic` more consistent with `assert_eq`.

I am not sure whether printing the `Any::type_id()` is useful, is there something better that we could print for non-string panic values?

### CODE REVIEW

File: src/libtest/test_result.rs [hunk 1]
```diff
@@ -49,9 +51,19 @@ pub fn calc_result<'a>(
                 if desc.allow_fail {
```
Comment (Centril) [inline src/libtest/test_result.rs]: Could be simplified to `else if desc.allow_fail` and `else if let Some(panic_str) = maybe_panic_str`

Comment (thomasetter) [reply src/libtest/test_result.rs]: Thanks, looks much better and will look nicer once allow_fail is removed too

**Comment by joshtriplett** [comment]
Looks reasonable to me!

@bors r+

**Comment by hdhoang** [comment]
@bors retry
chocolatey install timeout

**Pull request merged**